### ISSUE TITLE
Wheel failing to build in Docker container: `#include "frog/Frog.h"`

### ISSUE THREAD

**Issue by a3lem** [proycon/python-frog#6]
I'm trying to set up Frog 'from scratch' in a more pared-down Docker image (LaMachine is great, but at 3.76GB it's by far our largest image). While Frog seems to build just fine itself (I've included my Dockerfile below), `pip install python-frog` is still complaining that it can't find `frog/Frog.h`.

Here's the complete error log:

```
collecting python-frog
  Using cached python-frog-0.3.3.tar.gz
Building wheels for collected packages: python-frog
  Running setup.py bdist_wheel for python-frog ... error
  Complete output from command /usr/local/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-hqzo6jl_/python-frog/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/tmp61scgxewpip-wheel- --python-tag cp36:
  /usr/local/lib/python3.6/distutils/extension.py:131: UserWarning: Unknown Extension options: 'pyrex_gdb'
    warnings.warn(msg)
  running bdist_wheel
  running build
  running build_ext
  cythoning frog_wrapper.pyx to frog_wrapper.cpp
  building 'frog' extension
  creating build
  creating build/temp.linux-x86_64-3.6
  gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/root/local/include/ -I/usr/include/ -I/usr/include/libxml2 -I/usr/local/include/ -I/usr/local/include/python3.6m -c frog_wrapper.cpp -o build/temp.linux-x86_64-3.6/frog_wrapper.o --std=c++0x
  cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
  frog_wrapper.cpp:550:23: fatal error: frog/Frog.h: No such file or directory
   #include "frog/Frog.h"
                         ^
  compilation terminated.
  error: command 'gcc' failed with exit status 1

  ----------------------------------------
  Failed building wheel for python-frog
  Running setup.py clean for python-frog
Failed to build python-frog
Installing collected packages: python-frog
  Running setup.py install for python-frog ... error
    Complete output from command /usr/local/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-hqzo6jl_/python-frog/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-v38ixlqg-record/install-record.txt --single-version-externally-managed --compile:
    /usr/local/lib/python3.6/distutils/extension.py:131: UserWarning: Unknown Extension options: 'pyrex_gdb'
      warnings.warn(msg)
    running install
    running build
    running build_ext
    skipping 'frog_wrapper.cpp' Cython extension (up-to-date)
    building 'frog' extension
    creating build
    creating build/temp.linux-x86_64-3.6
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -I/root/local/include/ -I/usr/include/ -I/usr/include/libxml2 -I/usr/local/include/ -I/usr/local/include/python3.6m -c frog_wrapper.cpp -o build/temp.linux-x86_64-3.6/frog_wrapper.o --std=c++0x
    cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
    frog_wrapper.cpp:550:23: fatal error: frog/Frog.h: No such file or directory
     #include "frog/Frog.h"
                           ^
    compilation terminated.
    error: command 'gcc' failed with exit status 1

    ----------------------------------------
Command "/usr/local/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-hqzo6jl_/python-frog/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-v38ixlqg-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-hqzo6jl_/python-frog/
```

And here's my Dockerfile, just in case I'm still missing a dependency (pretty sure there's some redundancy with all the cython installations). The parent image is Debian-based, and gcc is present.

```
FROM python:latest

RUN apt-get update && \
    apt-get install -y libfolia-dev \
                       libticcutils2-dev \
                       ucto \
                       timbl \
                       timblserver \
                       mbt \
                       libicu-dev \
                       libxml2-dev \
                       frog \
                       frogdata \
                       python-dev \
                       build-essential \
                       cython3 \
                       cython 

RUN pip install cython \ 
                python-frog \
```

Any chance you can tell me what I should be doing differently here? 

Thanks in advance!

**Comment by proycon** [comment]
Hmm, it looks like you are indeed missing a dependency: ``libfrog-dev``... but the situation looks worse even: that package seems to be missing from debian entirely! It seems we overlooked it in the packaging process unfortunately :( Sorry.

I have made note of it (LanguageMachines/frog#42) and hope to do another packaging round soonish, in which I will correct the error. But debian being debian it will of course take a while to land in the distribution (it will first land in debian testing, not sure if that's an option for your container)

For now your best workaround, if you don't want to use LaMachine, is to compile Frog itself from scratch, or perhaps rely on Arch Linux instead of Debian and grab the package from the AUR (like LaMchine does).

**Comment by a3lem** [comment]
Haha, whoops! That is indeed unfortunate.

For the sake of convenience I'm currently using the official Python image as my base, which doesn't appear to come in *testing* flavors (which is probably just as well, seeing as our containers are intended for production). Still, a one-off switch to *testing* should be easy enough to do once the package lands. Arch I'm slightly more reluctant about, at least for the moment, most of our infrastructure being Debian/Ubuntu-based.

So, it looks like Iike I'm going to be having a crack at compiling Frog from scratch. Question: does that mean I will first have to compile `libfrog-dev` as well, or is that an artefact of compiling Frog? (I have little to no experience with C++, so apologies for the possibly stupid question.)

In addition, It might be worth observing that `libfrog-dev` is not mentioned anywhere on the [frog](https://github.com/LanguageMachines/frog) or [python-frog](https://github.com/proycon/python-frog) project pages, nor in the PDF manual. Would you perhaps consider appending a note to the README?

Anyway, thanks for all your work, especially on the LaMachine Docker image! It's the only way I've gotten any of the LanguageMachines tools to work, but so far the results have really been very impressive.

**Comment by a3lem** [comment]
Update: I just gone done trying to get _python-frog_ to work with a locally compiled version of Frog. `pip install python-frog` threw the same error as before, but `python setup.py install` seemed to work (I didn't try this second approach in my previous attempt), except for this warning:

```
Unknown Extension options: 'pyrex_gdb'warnings.warn(msg).
```  

Anyway,  I'm now getting an error when I try to `import frog`:

```
>>> import frog
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: libfolia.so.2: cannot open shared object file: No such file or directory```
```

(Earlier, I got a different error message, also relating to _libfolia_, about some unresolved symbol. It's fallen off my terminal's history, unfortunately, so I can't provide it here.)

Here's the relevant output of `ldconfig -v`:

```
/usr/local/lib:
        libpython3.6m.so.1.0 -> libpython3.6m.so.1.0
        libpython3.so -> libpython3.so
        libucto.so.2 -> libucto.so.2.0.0
        libfrog.so.1 -> libfrog.so.1.0.0
        libmbt.so.1 -> libmbt.so.1.0.0
        libtimbl.so.4 -> libtimbl.so.4.0.1
        libfolia.so.6 -> libfolia.so.6.0.3
        libticcutils.so.2 -> libticcutils.so.2.0.0
```

In order to compile *frog*, I stepped through `./configure`, resolving 'missing package' errors one by one. In total, I cloned and compiled the following packages (in the order that `./configure` complained about them).

1. ticcutils
2. timbl
3. mbt
4. folia
5. ucto
6. uctodata
7. frogdata

It looks like I'm going to have to start considering the Arch route, which, as I've explained previously, is less than ideal. 

Without coming across as ungrateful, we are starting to wish that the kind people at Radboud University's CLST would put some more thought into packaging their tools for easy installation. 12+ hours to set up a PoS tagger is a difficult sell in an enterprise setting. Again, though, we are very grateful to you for making them available.

**Comment by proycon** [comment]
> For the sake of convenience I'm currently using the official Python image as my base, which doesn't appear to come in testing flavors

Ah, I see you're referring to whatever Docker ships for ``python:latest`` right. I was actually referring to the Debian distribution, which comes in ``stable`` and ``testing`` flavours. I suspect your docker image is not based on the latest debian stable, based on some of the info you provided. 

> Arch I'm slightly more reluctant about, at least for the moment, most of our infrastructure being Debian/Ubuntu-based.

Yeah, fair enough. No need to go to Arch; Debian/Ubuntu is a a valid option which we fully support, especially for production. It was just a possible shortcut considering the missing ``frog-dev`` issue. I see you actually went the hard-core manual way compiling everything from scratch even :)
 
> In order to compile frog, I stepped through ./configure, resolving 'missing package' errors one by one. In total, I cloned and compiled the following packages (in the order that ./configure complained about them).
> 
>     ticcutils
>     timbl
>     mbt
>     folia
>     ucto
>     uctodata
>     frogdata

Perfect, that's a right order indeed.

> except for this warning:
> 
> Unknown Extension options: 'pyrex_gdb'warnings.warn(msg).

You can just ignore that one.

> Anyway, I'm now getting an error when I try to import frog:
> 
> >>> import frog
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ImportError: libfolia.so.2: cannot open shared object file: No such file or directory```

Okay. This suggests python-frog was compiled against ``libfolia.so.2``, quite an old version, rather than the new ``libfolia.so.6``. I guess this older libfolia library was lingering on the system at compile time?
Do you still have our debian packages installed? You'll want to remove all our packages before you compile from scratch otherwise you get such nasty version conflicts. Also, judging from the ``libfolia.so.2``; I suspect you might not have been running the latest Debian stable ('stretch' aka Debian 9)? That would mean the packages you tried to install earlier were pretty outdated anyhow.

So to fix this immediately there are two options:
* *[requires debian 9]*: Install all our debian packages (including ``-dev`` versions) except for frog/libfrog. Compile frog from source.
* *[would also work on debian 8]*: Do not install any of our packages (make sure they are all removed!), and recompile/install everything from source (If you pull from github I recommend using the latest releases rather than the master branch, which is by definition experimental).
* *[requires debian 9]*: Wait for a bit until I have the new packages ready; I can provide the debs manually so you can already try them even if they haven't hit the debian servers yet (that takes a while always, and as said, first they hit only testing).

> Without coming across as ungrateful, we are starting to wish that the kind people at Radboud University's CLST would put some more thought into packaging their tools for easy installation. 12+ hours to set up a PoS tagger is a difficult sell in an enterprise setting. Again, though, we are very grateful to you for making them available.

Actually, we spend quite a lot of time on packaging/distribution, because I totally agree with you that it indeed shouldn't be so difficult. That's first of all why we provide LaMachine (in three different flavours) and have packages for Debian and Arch Linux.  In this case the issue is more that we forgot to package ``libfrog-dev``, needed for ``python-frog``, an unfortunate mistake on our part which leads to the current trouble unfortunately..  But all in all I dare even say the amount of packaging effort is rather unusually high for a relatively small group such as ourselves ;)

**Comment by a3lem** [comment]
Hi @proycon, sorry for taking so long to get back to this! 

The plan was for this reply to include a working Dockerfile that handles the setup of `python-frog`, `frog` and all dependencies from scratch. And while I'm happy to report that I *was* eventually able to get everything installed manually inside a running Debian 9 'stretch' container, I never found the time to codify the steps in a Dockerfile. So for the past few weeks I've just been using the LaMachine image. 

On the plus side, Frog has been serving us very well. Based on what I've seen while going over its output, I daresay it may be the best PoS-tagger for Dutch right now (a proper comparison with other options is still in the works), and its NER does quite well, too.

As for my remark about ease of installation, I'd again like to say thanks for all the work you guys have put into all the tools in the LanguageMachine suite. I guess what I was driving at before stemmed from my frustration with the lack of a 'one-click' sort of solution. (As a Pythonista, perhaps I'm a bit spoiled. =p )

Anyway, I think this issue can be closed. I'm going to try to have another crack at things in a couple of months. In the meantime, see you at CLIN! :)

**Comment by proycon** [comment]
Glad you managed to sort it out! If you're doing a comparative study of Frog with other alternatives (regarding PoS and/or other annotations), then that would be very welcome and very interesting for  publication I'd say! It's been a long time since such studies were done (if at all;  so they're very out of date).

**Comment by frdg** [comment]
Hello!

Sorry to reopen this ticket, but in case that helps, here is a little recipe to get Frog running on Ubuntu.

Download and compile the following dependencies **in that specific order**:
```
ticcutils-0.17
libfolia-dev (from repo)
ucto-0.11
timbl-6.4.11
timblserver-1.11
mbt-3.3.1
libicu-dev (from repo)
libxml2-dev (from repo)
frogdata-0.15
```

Move to the dependency's folder, then:
```
bash bootstrap.sh
./configure
make
sudo make install
```

Once done, compile frog-0.13.9 and let the happiness flow through your veins.

**Comment by a3lem** [comment]
Thanks for taking the time to write that up, @frdg! Should definitely come in handy on my next attempt.

**Comment by emanjavacas** [comment]
Sorry for reopening, but I am having the same issue even though frog is available in the machine.

```bash
$ frog --version
frog 0.12.17 (c) ILK 1998 - 2013
Induction of Linguistic Knowledge Research Group, Tilburg University
based on [ucto 0.5.3, libfolia 0.10, timbl 6.4.4, timblserver 1.7, mbt 3.2.10]
```

but...

```bash
$ /usr/bin/python3.4 setup.py install
/usr/lib/python3.4/distutils/extension.py:132: UserWarning: Unknown Extension options: 'pyrex_gdb'
  warnings.warn(msg)
/usr/lib/python3.4/distutils/dist.py:260: UserWarning: Unknown distribution option: 'install_requires'
  warnings.warn(msg)
running install
running build
running build_ext
cythoning frog_wrapper.pyx to frog_wrapper.cpp
building 'frog' extension
creating build
creating build/temp.linux-x86_64-3.4
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -g -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/ -I/usr/include/libxml2 -I/usr/local/include/ -I/usr/include/python3.4m -c frog_wrapper.cpp -o build/temp.linux-x86_64-3.4/frog_wrapper.o --std=c++0x -D U_USING_ICU_NAMESPACE=1
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++ [enabled by default]
frog_wrapper.cpp:285:28: fatal error: frog/Frog-util.h: No such file or directory
 #include "frog/Frog-util.h"
                            ^
compilation terminated.
error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
```
 I have `libfrog.so.1` and `libfrog.so.1.0.0` inside `/usr/lib/`.

**Comment by proycon** [comment]
Are you using the debian packages? The problem with the missing ``libfrog-dev`` is unfortunately still not resolved  (LanguageMachines/Frog#42), which would explain your error.

Your easiest solution is simply to use LaMachine (https://proycon.github.io/LaMachine), in whatever form you prefer, python-frog is included there..  alternatively you can compile everything from scratch as indicated by @frdg .

**Comment by proycon** [comment]
(Closing this issue because it is old and should be resolved in the meantime, the missing libfrog-dev should also be available in the latest debian, although by now it is already outdated again)
